// astro/dist/core/dev/dev.js
import { performance } from "perf_hooks";
import { attachContentServerListeners } from "../../content/index.js";
import { info, warn } from "../logger/core.js";
import * as msg from "../messages.js";
import { startContainer } from "./container.js";
import { createContainerWithAutomaticRestart } from "./restart.js";
async function dev(settings, options) {
  var _a, _b;
  const devStart = performance.now();
  await options.telemetry.record([]);
  const restart = await createContainerWithAutomaticRestart({
    flags: options.flags ?? {},
    handleConfigError: options.handleConfigError,
    beforeRestart: () => console.clear(),
    params: {
      settings,
      logging: options.logging,
      isRestart: options.isRestart
    }
  });
  const devServerAddressInfo = await startContainer(restart.container);
  info(
    options.logging,
    null,
    msg.serverStart({
      startupTime: performance.now() - devStart,
      resolvedUrls: restart.container.viteServer.resolvedUrls || { local: [], network: [] },
      host: settings.config.server.host,
      base: settings.config.base,
      isRestart: options.isRestart
    })
  );
  const currentVersion = "2.0.10";
  if (currentVersion.includes("-")) {
    warn(options.logging, null, msg.prerelease({ currentVersion }));
  }
  if (((_b = (_a = restart.container.viteConfig.server) == null ? void 0 : _a.fs) == null ? void 0 : _b.strict) === false) {
    warn(options.logging, null, msg.fsStrictWarning());
  }
  await attachContentServerListeners(restart.container);
  return {
    address: devServerAddressInfo,
    get watcher() {
      return restart.container.viteServer.watcher;
    },
    handle(req, res) {
      return restart.container.handle(req, res);
    },
    async stop() {
      await restart.container.close();
    }
  };
}
export {
  dev as default
};
