// astro/dist/vite-plugin-head-propagation/index.js
import { walkParentInfos } from "../core/build/graph.js";
import { getAstroMetadata } from "../vite-plugin-astro/index.js";
const injectExp = /^\/\/\s*astro-head-inject/;
function configHeadPropagationVitePlugin({
  settings
}) {
  function addHeadInjectionInTree(graph, id, getInfo, seen = /* @__PURE__ */ new Set()) {
    const mod = server.moduleGraph.getModuleById(id);
    for (const parent of (mod == null ? void 0 : mod.importers) || []) {
      if (parent.id) {
        if (seen.has(parent.id)) {
          continue;
        } else {
          seen.add(parent.id);
        }
        const info = getInfo(parent.id);
        if (info == null ? void 0 : info.meta.astro) {
          const astroMetadata = getAstroMetadata(info);
          if (astroMetadata) {
            astroMetadata.propagation = "in-tree";
          }
        }
        addHeadInjectionInTree(graph, parent.id, getInfo, seen);
      }
    }
  }
  let server;
  return {
    name: "astro:head-propagation",
    configureServer(_server) {
      server = _server;
    },
    transform(source, id) {
      if (!server) {
        return;
      }
      if (injectExp.test(source)) {
        addHeadInjectionInTree(server.moduleGraph, id, (child) => this.getModuleInfo(child));
      }
    }
  };
}
function astroHeadPropagationBuildPlugin(options, internals) {
  return {
    build: "ssr",
    hooks: {
      "build:before"() {
        const map = /* @__PURE__ */ new Map();
        return {
          vitePlugin: {
            name: "vite-plugin-head-propagation-build",
            generateBundle(_opts, bundle) {
              const appendPropagation = (info) => {
                const astroMetadata = getAstroMetadata(info);
                if (astroMetadata) {
                  astroMetadata.propagation = "in-tree";
                  map.set(info.id, "in-tree");
                }
              };
              for (const [bundleId, output] of Object.entries(bundle)) {
                if (output.type !== "chunk")
                  continue;
                for (const [id, mod] of Object.entries(output.modules)) {
                  if (mod.code && injectExp.test(mod.code)) {
                    for (const [info2] of walkParentInfos(id, this)) {
                      appendPropagation(info2);
                    }
                  }
                  const info = this.getModuleInfo(id);
                  if (info) {
                    appendPropagation(info);
                  }
                }
              }
              internals.propagation = map;
            }
          }
        };
      }
    }
  };
}
export {
  astroHeadPropagationBuildPlugin,
  configHeadPropagationVitePlugin as default
};
